### PR TITLE
On 276 - Including GivenNames and FamilyName to Submission Events

### PR DESCRIPTION
**Pull request by davidgouge**
Also always including CommitmentId in SubmissionEvents

In SubmissionEventsProcessor, I've also included StandardCode and ActualStartDate in the 'always include in SubmissionEvents' section. Would this cause a problem for anyone else?  They are fields that we need returning to us every time.

### CODE REVIEW

**Comment by davidgouge** [comment]
Agh, fixing up failing test now.

File: src/SharedPipelineComponents/IlrSubmissionEvents/DeploySubmission/sql/migrationscripts/~2_Submissions.Deds.AddColumns.sql [hunk 1]
```diff
@@ -14,4 +14,37 @@ IF NOT EXISTS(SELECT [column_id] FROM sys.columns WHERE [name] = 'EPAOrgId' AND
     ALTER TABLE [Submissions].[SubmissionEvents]
 	ADD [EPAOrgId] varchar(7) NULL
   END
+GO
+
```
Comment (MichaelYoung1981) [inline src/SharedPipelineComponents/IlrSubmissionEvents/DeploySubmission/sql/migrationscripts/~2_Submissions.Deds.AddColumns.sql]: You'll need to move this into a new migration script as number 2 will already have ran.

**Review by MichaelYoung1981 — DISMISSED** [review]

File: src/SharedPipelineComponents/IlrSubmissionEvents/DeploySubmission/copy mappings/DasSubmissionEventsCopyToDedsMapping.xml [hunk 2]
```diff
@@ -105,6 +105,14 @@
...
           <Destination>EPAOrgId</Destination>
         </ColumnMapping>
+        <ColumnMapping>
+          <Source>GivenNames</Source>
+          <Destination>GivenNames</Destination>
+        </ColumnMapping>
+        <ColumnMapping>
+          <Source>FirstName</Source>
+          <Destination>FirstName</Destination>
+        </ColumnMapping>
```
Comment (carlbm) [inline src/SharedPipelineComponents/IlrSubmissionEvents/DeploySubmission/copy mappings/DasSubmissionEventsCopyToDedsMapping.xml]: GivenNames and FirstName? Should one of these be surname?

Comment (carlbm) [reply src/SharedPipelineComponents/IlrSubmissionEvents/DeploySubmission/copy mappings/DasSubmissionEventsCopyToDedsMapping.xml]: Or FamilyName

File: src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission/SubmissionEventsProcessor.cs [hunk 3]
```diff
@@ -171,6 +181,12 @@ private SubmissionEvent CalculateDelta(IlrDetails currentIlr, IlrDetails lastSee
                 // EPAOrgId is optional in the ilr, so we need to always set it, otherwise if it is null,
                 // the consumer won't know if it hasn't changed or if it's been removed on a subsequent irl submission
                 @event.EPAOrgId = currentIlr.EPAOrgId;
+                @event.GivenNames = currentIlr.GivenNames;
+                @event.FamilyName = currentIlr.FamilyName;
+                @event.CommitmentId = currentIlr.CommitmentId;
+
+                @event.StandardCode = currentIlr.StandardCode;
+                @event.ActualStartDate = currentIlr.ActualStartDate;
```
Comment (carlbm) [inline src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission/SubmissionEventsProcessor.cs]: You've added these fields that don't have anything to do with given/family names. Was there a reason?

Comment (carlbm) [reply src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission/SubmissionEventsProcessor.cs]: Just saw your commit comment. I'm not sure if these changes will have an impact. I'll speak to Mike about it

**Review by carlbm — CHANGES_REQUESTED** [review]
FirstName to FamilyName in the xml files is a must for approval

File: src/SharedPipelineComponents/IlrSubmissionEvents/DeploySubmission/copy mappings/DasSubmissionEventsCopyToDedsMapping.xml (hunk 2 above)
Comment (davidgouge) [reply src/SharedPipelineComponents/IlrSubmissionEvents/DeploySubmission/copy mappings/DasSubmissionEventsCopyToDedsMapping.xml]: Yes, of course, sorry.  Pushed a commit to change below.

**Review by carlbm — APPROVED** [review]

**Comment by MichaelYoung1981** [comment]
Any idea when this is going to be merged/deployed?

File: src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission.UnitTests/SubmissionEventsProcessorTests/WhenProcessing.cs [hunk 4]
```diff
@@ -148,13 +159,15 @@ public void ThenItShouldWriteAnEventForEachIlrDeliveryThatHasChanged()
...
+            _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].ActualStartDate == updatedSubmissionChanged.ActualStartDate)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].PlannedEndDate == updatedSubmissionChanged.PlannedEndDate)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].ActualEndDate == null)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].OnProgrammeTotalPrice == updatedSubmissionChanged.OnProgrammeTotalPrice)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].CompletionTotalPrice == updatedSubmissionChanged.CompletionTotalPrice)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].NiNumber == null)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[0].EPAOrgId == ilrForFirstSubmission.EPAOrgId)));
+            _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[0].GivenNames == ilrForFirstSubmission.GivenNames)));
```
Comment (lazcool) [inline src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission.UnitTests/SubmissionEventsProcessorTests/WhenProcessing.cs]: The last 3 lines should be checking c.Events[1]

Comment (davidgouge) [reply src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission.UnitTests/SubmissionEventsProcessorTests/WhenProcessing.cs]: 😡 Damn it!  Fixing up now.  👍

Comment (davidgouge) [reply src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission.UnitTests/SubmissionEventsProcessorTests/WhenProcessing.cs]: Fixed below.

**Comment by lazcool** [comment]
Always including StandardCode and ActualStartDate in the submission event shouldn't cause any problems. In fact, we should probably include all fields in all submission events and remove the idea of only storing the delta.
This is especially true for optional fields, as it's impossible to tell the difference between the field being changed to null and there being no change to the field!

**Comment by MichaelYoung1981** [comment]
I agree with @lazcool, it'd simplify a lot and given we know how the consumers handle the events it wouldn't cause any issues (that I can think of!).

**Comment by chrisfoster76** [comment]
@davidgouge this branch is currently conflicted, could you please update it?

File: src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission.UnitTests/SubmissionEventsProcessorTests/WhenProcessing.cs [hunk 5]
```diff
@@ -156,12 +171,16 @@ public void ThenItShouldWriteAnEventForEachIlrDeliveryThatHasChanged()
...
-            _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].ActualStartDate == null)));
+            _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].ActualStartDate == updatedSubmissionChanged.ActualStartDate)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].PlannedEndDate == updatedSubmissionChanged.PlannedEndDate)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].ActualEndDate == null)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].OnProgrammeTotalPrice == updatedSubmissionChanged.OnProgrammeTotalPrice)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].CompletionTotalPrice == updatedSubmissionChanged.CompletionTotalPrice)));
             _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].NiNumber == null)));
+            _mediator.Verify(m => m.Send(It.Is<WriteSubmissionEventCommand>(c => c.Events[1].EPAOrgId == ilrForFirstSubmission.EPAOrgId)));
```
Comment (lazcool) [inline src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission.UnitTests/SubmissionEventsProcessorTests/WhenProcessing.cs]: This check already exists below.

Comment (davidgouge) [reply src/SharedPipelineComponents/IlrSubmissionEvents/SFA.DAS.Provider.Events.Submission.UnitTests/SubmissionEventsProcessorTests/WhenProcessing.cs]: Yeah, looks like a merge issue.  Fixed. 👍

**Review by lazcool — CHANGES_REQUESTED** [review]
I think the new fields need to be added to SubmissionEvent's ilr.transient.ddl.tables.sql...
CompStatus into [Valid].[LearningDelivery] 
FamilyName and GivenNames into [Valid].[Learner]

and into the insert statement in 01 submissions.populate.submissions.sql

and these integration tests need to be updated to take account of the new fields on SubmissionEvents & LastSeenVersion...
WhenANewIlrFileIsSubmitted.ThenItShouldWriteANewSubmissionEventWithAllPropertiesPopulated()
WhenANewIlrFileIsSubmitted.ThenItShouldWriteIlrDetailsToLastSeenVersion()
WhenAnIlrFileIsSubmittedWithNoSignificantUpdates.ThenItShouldWriteIlrDetailsToLastSeenVersion()
WhenAnUpdatedIlrFileIsSubmitted.ThenItShouldWriteANewSubmissionEventWithOnlyUpdatedAndStandardPropertiesPopulated()

WhenPreparingTransientDatabaseForCollection.ThenItShouldCopyReferenceDataFromDedsToTransient()

**Comment by carlbm** [comment]
Partly agree with @lazcool, but we don't own the Valid scheme, so we can't make any changes to those tables.

The deds tables don't have the columns, so it will fail at the copy to deds stage. The fields will need to be added via a migration script

The DML doesn't reference the new columns (firstname, surname, compstatus), so will always be blank. I see that there's a view, but I couldn't find a reference to the view. Looks like the repo is looking directly at the Submissions.* tables

**Review by carlbm — CHANGES_REQUESTED** [review]
As per my comments

**Comment by davidgouge** [comment]
Integration tests updated to include new fields.

**Comment by davidgouge** [comment]
@carlbm I'm not sure what most of that means.  I've made changes in all (I think, now) the same places where the EPAOrgId was added.  All ILR Submission Integration tests now pass.

**Review by lazcool — APPROVED** [review]
The comments I raised have been addressed.
I don't know enough about changing the Valid schema / migration scripts, so I'll let @carlbm approve that bit.

**Comment by davidgouge** [comment]
Updated migration files as per @carlbm's instruction.

**Review by carlbm — APPROVED** [review]

**Comment by satya-sanikommu** [comment]
Review and Testing completed. Ready to merge

**Pull request merged**